query/src/cmd/compaction.rs

```diff
@@ -0,0 +1,255 @@
+// Copyright 2021 Datafuse Labs.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::HashMap;
+use std::sync::Arc;
+use std::task::Context;
+use std::task::Poll;
+use std::time::Duration;
+
+use common_base::base::tokio;
+use common_base::base::tokio::sync::mpsc::channel;
+use common_base::base::tokio::sync::mpsc::error::SendError;
+use common_base::base::tokio::sync::mpsc::Receiver;
+use common_base::base::tokio::sync::mpsc::Sender;
+use common_base::base::TrySpawn;
+use common_exception::ErrorCode;
+use common_meta_api::deserialize_struct;
+use common_meta_api::get_start_and_end_of_prefix;
+use common_meta_api::KVApiKey;
+use common_meta_api::PREFIX_TABLE_BY_ID;
+use common_meta_app::schema::TableId;
+use common_meta_app::schema::TableIdent;
+use common_meta_app::schema::TableInfo;
+use common_meta_app::schema::TableMeta;
+use common_meta_grpc::MetaGrpcClient;
+use common_meta_types::protobuf::watch_request::FilterType;
+use common_meta_types::protobuf::WatchRequest;
+use common_planners::OptimizeTableAction;
+use common_planners::OptimizeTablePlan;
+use futures::ready;
+use futures::Stream;
+use tokio_stream::StreamExt;
+use tokio_util::time::delay_queue;
+use tokio_util::time::DelayQueue;
+
+use crate::catalogs::CATALOG_DEFAULT;
+use crate::sessions::QueryContext;
+use crate::sessions::TableContext;
+use crate::storages::Table;
+use crate::Config;
+
+pub async fn compaction(ctx: Arc<QueryContext>, conf: &Config) -> common_exception::Result<()> {
+    let rpc_conf = conf.meta.to_meta_grpc_client_conf();
+    if rpc_conf.local_mode() {
+        return Err(ErrorCode::CompactionError(
+            "Auto compaction is not supported in local mode".to_string(),
+        ));
+    }
+
+    let (key, key_end) = get_start_and_end_of_prefix(PREFIX_TABLE_BY_ID)?;
+    let watch = WatchRequest {
+        key,
+        key_end: Some(key_end),
+        filter_type: FilterType::Update.into(),
+    };
+    // Create a new client of metasrv.
+    let client = MetaGrpcClient::try_new(&rpc_conf)?;
+
+    // Watch the table meta update.
+    let mut client_stream = client.request(watch).await?;
+    let task = CompactionQueue::new(Duration::from_secs(3600));
```

Comment:
Maybe we can make `timeout` as a user-specified argument, not a hardcode number?

Reply:
As discussed in https://github.com/datafuselabs/databend/pull/6508#issuecomment-1189712979

I don't think place auto-compaction in command line args is a good UX.

Reply:
> Maybe we can make `timeout` as a user-specified argument, not a hardcode number?

We will consider making `timeout` and `cron time` as user-specified arguments later.